src/USER-MISC/fix_charge_regulation.cpp

```diff
@@ -0,0 +1,1335 @@
+      count = count + 1;
+  }
+  int count_sum = count;
+  MPI_Allreduce(&count, &count_sum, 1, MPI_INT, MPI_SUM, world);
+  return count_sum;
+}
+
+double Fix_charge_regulation::compute_vector(int n) {
+  double count_temp = 0;
+  if (n == 0) {
+    return nacid_attempts + nbase_attempts + nsalt_attempts;
+  } else if (n == 1) {
+    return nacid_successes + nbase_successes + nsalt_successes;
+  } else if (n == 2) {
+    return particle_number(acid_type, 0);
+  } else if (n == 3) {
+    return particle_number(acid_type, -1);
+  } else if (n == 4) {
+    return particle_number(base_type, 0);
+  } else if (n == 5) {
+    return particle_number(base_type, 1);
+  } else if (n == 6) {
+    return particle_number(cation_type, salt_charge[0]);
+  } else if (n == 7) {
+    return particle_number(anion_type, salt_charge[1]);
+  }
+  return 0.0;
+}
+
+void Fix_charge_regulation::setThermoTemperaturePointer() {
+  int ifix = -1;
+  ifix = modify->find_fix(idftemp);
+  if (ifix == -1) {
+    error->all(FLERR,
+               "Fix charge regulation regulation could not find a temperature fix id provided by tempfixid\n");
+  }
+  Fix *temperature_fix = modify->fix[ifix];
+  int dim;
+  target_temperature_tcp = (double *) temperature_fix->extract("t_target", dim);
+
+}
+
+void Fix_charge_regulation::assign_tags() {
+  // Assign tags to ions with zero tags
+  if (atom->tag_enable) {
+    tagint *tag = atom->tag;
+    tagint maxtag_all = 0;
+    tagint maxtag = 0;
+    for (int i = 0; i < atom->nlocal; i++) maxtag = MAX(maxtag, tag[i]);
+    maxtag_all = maxtag;
+    MPI_Allreduce(&maxtag, &maxtag_all, 1, MPI_LMP_TAGINT, MPI_MAX, world);
+    if (maxtag_all >= MAXTAGINT)
+      error->all(FLERR, "New atom IDs exceed maximum allowed ID");
+
+    tagint notag = 0;
+    tagint notag_all;
+    for (int i = 0; i < atom->nlocal; i++)
+      if (tag[i] == 0 && (atom->type[i] == cation_type || atom->type[i] == anion_type))notag++;
+    notag_all = notag;
+    MPI_Allreduce(&notag, &notag_all, 1, MPI_LMP_TAGINT, MPI_SUM, world);
+    if (notag_all >= MAXTAGINT)
+      error->all(FLERR, "New atom IDs exceed maximum allowed ID");
+
+    tagint notag_sum = notag;
+    MPI_Scan(&notag, &notag_sum, 1, MPI_LMP_TAGINT, MPI_SUM, world);
+    // itag = 1st new tag that my untagged atoms should use
+
+    tagint itag = maxtag_all + notag_sum - notag + 1;
+    for (int i = 0; i < atom->nlocal; i++) {
+      if (tag[i] == 0 && (atom->type[i] == cation_type || atom->type[i] == anion_type)) {
+        tag[i] = itag++;
+      }
+    }
+    if (atom->map_style) atom->map_init();
+    atom->nghost = 0;
+    comm->borders();
+  }
+}
+
+/* ----------------------------------------------------------------------
+   parse input options
+------------------------------------------------------------------------- */
+
+void Fix_charge_regulation::options(int narg, char **arg) {
+  if (narg < 0) error->all(FLERR, "Illegal fix charge regulation command");
+
+  // defaults
+
+  pH = 7.0;
```

Comment:
I see a lot of expressions like `pow(10,-pXXX)` or `pow(10,-pXXX+pYYY)` throughout the code. Since the `pow()` function is among the most expensive functions in the standard math library, one can sometimes achieve a significant performance boost by pre-computing and storing those values in stead of the "natural" ones and thus avoid having to use `pow()` a lot during the evaluation of the body of this fix but either use the pre-computed value directly or product/fraction in case exponents are added or subtracted.